File: week_05_ basic_key_exchange/src/attempts/DiscreteLogModulo.java
package attempts;

/** javac -d bin -sourcepath src src/com/DiscreteLogModulo.java  */
/** java -cp bin/ com.DiscreteLogModulo */

import java.math.BigInteger;
import java.util.Map;
import java.util.HashMap;
import java.util.Date;

/** Calculate Discrete Log Modulo. */

/**
 * 
 BigInteger p = ""; BigInteger g ; BigInteger h;
 * 
 * 
 * 
 * h/(g^x)
 */
public class DiscreteLogModulo {

	//	public static final Integer B = (int)Math.pow(2, 5); 
	//	public static final String P_STR = "17"; 
	//	public static final String G_STR = "3"; 
	//	public static final String H_STR = "12";

//	 public static final Integer B = (int)Math.pow(2, 10);
//	 public static final String P_STR = "110213";
//	 public static final String G_STR = "101033";
//	 public static final String H_STR = "100551";

	public static final Integer B = (int) Math.pow(2, 20);
	public static final String P_STR = "13407807929942597099574024998205846127479365820592393377723561443721764030073546976801874298166903427690031858186486050853753882811946569946433649006084171";
	public static final String G_STR = "11717829880366207009516117596335367088558084999998952205599979459063929499736583746670572176471460312928594829675428279466566527115212748467589894601965568";
	public static final String H_STR = "3239475104050450443565264378728065788649097520952449527834792452971981976143292558073856937958553180532878928001494706097394108577585732452307673444020333";

	public static void main(String[] args) {
		Map<BigInteger, Integer> myMap = new HashMap<BigInteger, Integer>();
		BigInteger p = new BigInteger(P_STR);
		BigInteger g = new BigInteger(G_STR);
		BigInteger h = new BigInteger(H_STR);
		BigInteger x;
		BigInteger temp;
		BigInteger function = g.modInverse(p);
		BigInteger anterior = BigInteger.valueOf(1);
		anterior = anterior.modInverse(p);
		anterior = anterior.multiply(h).mod(p);
		myMap.put(anterior, 0);
		Date begin = new Date();
		for (int x1 = 1; x1 < B; x1++) {
			temp = anterior.multiply(function).mod(p);
			myMap.put(temp, x1);
			anterior = temp;
			if (x1 % 1000 == 0) {
				Date end = new Date();
				long diff = (end.getTime() - begin.getTime()) / 1000;
				System.out.print("primer bucle i=" + x1);
				System.out.println("\t secs: " + diff);
				//System.out.println("temp size: " + temp.toString().length());
				begin = new Date();
			}
		}
		long solution = -1;
		for (int x0 = 0; x0 < B && solution == -1; x0++) {
			// long exponent = B*i;
			// temp = g.modPow(BigInteger.valueOf(exponent), p);
			temp = g.modPow(BigInteger.valueOf(B), p);
			temp = temp.modPow(BigInteger.valueOf(x0), p);
			// System.out.println(i + "\tRHS " + temp);
			Integer x1 = myMap.get(temp);
			if (x1 != null) {
				System.out.println("middle " + temp);
				System.out.println("x0 " + x0);
				System.out.println("x1 " + x1);
				solution = x0 * B + x1;
			}
			if (x0 % 1000 == 0) {
				Date end = new Date();
				long diff = (end.getTime() - begin.getTime()) / 1000;
				System.out.print("segundo bucle i=" + x0);
				System.out.println("\t secs: " + diff);
				begin = new Date();
			}
		}
		System.out.println("finito: " + solution);
		/*
		 * for (int i=0; i<B; i++) { temp = g.pow(i); temp = temp.modInverse(p);
		 * temp = temp.multiply(h).mod(p); myMap.put(temp, i); if (i%1000==0) {
		 * Date end = new Date(); long diff = (end.getTime() - begin.getTime())
		 * / 1000; System.out.print ("primer bucle i=" + i); System.out.println
		 * ("\t secs: " + diff); System.out.println("temp size: " +
		 * temp.toString().length()); begin = new Date(); } } for (int i=0; i<B;
		 * i++) { temp = g.modPow(BigInteger.valueOf(B), p); temp =
		 * temp.modPow(BigInteger.valueOf(i), p); System.out.println(i +
		 * "\tRHS " + temp); }
		 */
	}

	public static void mainOls(String[] args) {
		BigInteger p = new BigInteger(P_STR);
		BigInteger g = new BigInteger(G_STR);
		BigInteger h = new BigInteger(H_STR);
		BigInteger x;

		Map<BigInteger, Integer> myMap = new HashMap<BigInteger, Integer>();
		Date begin = new Date();
		x = h.divide(g.pow(0));
		for (int i = 0; i < B; i++) {
			myMap.put(x.mod(p), i);
			x = h.divide(g);
			if (i % 10 == 0) {
				Date end = new Date();
				long diff = (end.getTime() - begin.getTime()) / 1000;
				System.out.print("primer bucle i=" + i);
				System.out.println("\t secs: " + diff);
				begin = new Date();
			}
		}
		boolean encontrado = false;
		long solution = -1;
		for (int i = 0; i < B && !encontrado; i++) {
			long exponent = B * i;
			x = g.modPow(BigInteger.valueOf(exponent), p);
			Integer x1 = myMap.get(x);
			if (x1 != null) {
				System.out.println("x0 " + i);
				System.out.println("x1 " + x1);
				encontrado = true;
				solution = i * B + x1;
			}
			if (i % 1000 == 0) {
				Date end = new Date();
				long diff = (end.getTime() - begin.getTime()) / 1000;
				System.out.print("segundo bucle i=" + i);
				System.out.println("\t secs: " + diff);
				begin = new Date();
			}
		}
		System.out.println("solution: " + solution);
	}
}

File: week_05_ basic_key_exchange/src/DiscreteLogModulo.py
#!/usr/bin/python


n1 = '1340780792994259709957402499820584612747936582059239337772\
3561443721764030073546976801874298166903427690031858186486050853\
753882811946569946433649006084171'
n2 = '1171782988036620700951611759633536708855808499999895220559\
9979459063929499736583746670572176471460312928594829675428279466\
566527115212748467589894601965568'
n3 = '3239475104050450443565264378728065788649097520952449527834\
7924529719819761432925580738569379585531805328789280014947060973\
94108577585732452307673444020333'
B = pow(2,20)
'''

n1 = '110213'
n2 = '101033'
n3 = '100551'
B = pow(2,10)
'''
'''
n1 = '17'
n2 = '3'
n3 = '12'
B = pow(2,5)
'''
p,g,h = int(n1),int(n2),int(n3)





def mod_inverse(a,P):
    x, px, b = 0, 1, P
    while b: x, px, a, b = px - (a // b) * x, x, b, a % b
    return px % P

def mod_pow(a, b, P):
    result = 1
    while b > 0:
        b, r = divmod(b, 2)
        if r: result = (result * a) % P
        a = (a * a) % P
    return result

diccionario = {}
function = mod_inverse(g, p)
anterior = mod_inverse(1, p)
anterior = (anterior * h) % p
x1 = 0
diccionario[anterior] = x1

while x1 < B:
    x1 += 1
    temp = (anterior*function) % p
    # print "guarda temp=" + str(temp)
    diccionario[temp] = x1
    anterior = temp;
    # print "inverse g=" + str(g) + " p=" + str(p) + " modInverse=" + str(anterior)
    if x1 % 1000 == 0:
        print "pasa X1 -->" + str(x1)
print "termina X1 ----" + str(x1)

x0 = 0;
encontrado = -1;
solution = -1;
while x0 < B and encontrado==-1:
    #print "x0 ----" + str(x0)   
    x0 += 1
    temp = mod_pow(g, B, p)
    temp = mod_pow(temp, x0, p)
    # print "temp" + str(temp)
    default = -1
    encontrado = diccionario.get(temp, default)
    #print " ->" + str(default)
    if encontrado != -1:
        print "encontrado" + str(encontrado)
        solution = x0 * B + encontrado;
    if x0 % 1000 == 0:
        print "pasa X0 -->" + str(x0)

    '''
    temp = g.modPow(BigInteger.valueOf(B), p);
	temp = temp.modPow(BigInteger.valueOf(x0), p);
    '''
print "termina x0 ----" + str(x0)    
print "termina encontrado ----" + str(encontrado)
print "termina solution ----" + str(solution)